### PR TITLE
Dask disk backend 

### LINKED ISSUE #677: Investigate issues with dask dropping workers and completed work in memory on failed tasks when running in jobqueue mode.
Dask is currently dropping a bunch of workers whenever anything is being run in `jobqueue` mode on lilac. This is very frustrating because it means completed work is being lost, as part of this we can reduce the number of times completed work is re-attempted. Currently testing.

### PR DESCRIPTION
**Pull request by hmacdope**
## Description
Provide a brief description of the PR's purpose here.
Fixes #677 and large docking workflow issues

Adds an initial disk based backed for dask, so that large transfers can be avoided and instead work generated directly on the workers. 

## Developers certificate of origin
- [X] I certify that this contribution is covered by the MIT license as defined in our [LICENSE](https://github.com/choderalab/asapdiscovery/blob/main/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

### CODE REVIEW

File: asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py [hunk 1]
```diff
@@ -295,17 +300,19 @@ def large_scale_docking_workflow(inputs: LargeScaleDockingInputs):
     logger.info("Scoring docking results")
     scorer = MetaScorer(scorers=scorers)
 
-    if inputs.write_final_sdf:
-        logger.info("Writing final docked poses to SDF file")
-        write_ligands_to_multi_sdf(
-            output_dir / "docking_results.sdf", [r.posed_ligand for r in results]
-        )
+    # if inputs.write_final_sdf:
```
Comment (hmacdope) [inline asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py:303]: Need to fix this.

**Comment by hmacdope** [comment]
@JenkeScheen this should fix dask issues we have been having, haven't tested on lilac because it is down but should be much improved. 

@JenkeScheen @chrisiacovella @apayne97 @kaminow @ijpulidos could I grab a review from one or more of yall? This is kinda blocking project support working correctly on `main`

File: asapdiscovery-data/asapdiscovery/data/dask_utils.py [hunk 2]
```diff
@@ -1,26 +1,46 @@
...
+
+    IN_MEMORY = "in-memory"
+    DISK = "disk"
+
+
 def set_dask_config():
     cfg.set({"distributed.scheduler.worker-ttl": None})
     cfg.set({"distributed.admin.tick.limit": "4h"})
-    cfg.set({"distributed.scheduler.allowed-failures": 1})
+    cfg.set({"distributed.scheduler.allowed-failures": 2})
```
Comment (chrisiacovella) [inline asapdiscovery-data/asapdiscovery/data/dask_utils.py:36]: Why the change to 2 here?

**Review by chrisiacovella — APPROVED** [review]
This looks good.

File: asapdiscovery-data/asapdiscovery/data/dask_utils.py [hunk 3]
```diff
@@ -252,7 +301,11 @@ def from_cpu(cls, cpu: CPU = CPU.LT):
 
 
 def dask_cluster_from_type(
-    dask_type: DaskType, gpu: GPU = GPU.GTX1080TI, cpu: CPU = CPU.LT
+    dask_type: DaskType,
+    gpu: GPU = GPU.GTX1080TI,
+    cpu: CPU = CPU.LT,
+    local_threads_per_worker: int = 1,
+    silence_logs: int = logging.DEBUG,  # worst kwarg name but it is what it is
```
Comment (apayne97) [inline asapdiscovery-data/asapdiscovery/data/dask_utils.py:308]: LOL

File: asapdiscovery-data/asapdiscovery/data/dask_utils.py [hunk 4]
```diff
@@ -271,9 +324,36 @@ def dask_cluster_from_type(
...
     """
+    logger.info(f"Platform: {get_platform()}")
+    cpu_count = psutil.cpu_count()
+    logger.info(f"Logical CPU count: {cpu_count}")
+    physical_cpu_count = psutil.cpu_count(logical=False)
+    logger.info(f"Physical CPU count: {physical_cpu_count}")
+
     logger.info(f"Getting dask cluster of type {dask_type}")
+    logger.info(f"Dask log level: {silence_logs}")
+
```
Comment (apayne97) [inline asapdiscovery-data/asapdiscovery/data/dask_utils.py:335]: this is nice

File: asapdiscovery-docking/asapdiscovery/docking/docking_v2.py [hunk 5]
```diff
@@ -287,3 +296,35 @@ def __eq__(self, other: Any) -> bool:
...
+
+    Parameters
+    ----------
+    results : Union[list[DockingResult], list[Path]]
+        List of DockingResults or paths to DockingResult json files
+    backend : BackendType, optional
+        Backend to use, by default BackendType.IN_MEMORY
+
+
+    """
```
Comment (apayne97) [inline asapdiscovery-docking/asapdiscovery/docking/docking_v2.py:318]: update docstring!

File: asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py [hunk 6]
```diff
@@ -153,7 +158,7 @@ def large_scale_docking_workflow(inputs: LargeScaleDockingInputs):
             dask_cluster.scale(inputs.dask_cluster_n_workers)
 
         dask_client = Client(dask_cluster)
-        # dask_client.forward_logging() distributed vs dask_cuda versioning issue, see # #669
+        dask_client.forward_logging()  # distributed vs dask_cuda versioning issue, see # #669
```
Comment (apayne97) [inline asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py:161]: is this no longer an issue?

**Review by apayne97 — COMMENTED** [review]
I don't feel like I quite understand dask and all the issues here well enough to know if this is a good way to do it, but if it is working then maybe that's good enough!

File: asapdiscovery-docking/asapdiscovery/docking/openeye.py [hunk 7]
```diff
@@ -197,31 +199,35 @@ def _dock(
...
-                    Path(
-                        Path(output_dir) / set.unique_name() / "docking_result.json"
-                    ).exists()
-                )
+                and (docked_result_json_path.exists())
             ):
-                print(
+                logger.info(
                     f"Docking result for {set.unique_name()} already exists, reading from disk"
                 )
```
Comment (apayne97) [inline asapdiscovery-docking/asapdiscovery/docking/openeye.py:227]: this broke this test: https://github.com/choderalab/asapdiscovery/blob/49310f8e369c928ef416b29bcbe75522919a86e4/asapdiscovery-docking/asapdiscovery/docking/tests/test_docking_v2.py#L71-L72

because it was relying on the `out` print statement. This needs to be changed for the test to pass!

**Review by apayne97 — DISMISSED** [review]
Minor testing fix!

File: asapdiscovery-data/asapdiscovery/data/dask_utils.py (hunk 2 above)
Comment (hmacdope) [reply asapdiscovery-data/asapdiscovery/data/dask_utils.py:36]: Basically if the work is borked, leading to a killed worker and the loss of a lot of progress we don't try it 3 times, instead only 2. 1 is too few IMHO and 3 can be a bit slow.

File: asapdiscovery-docking/asapdiscovery/docking/openeye.py (hunk 7 above)
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/openeye.py:227]: Should be fixed now.

File: asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py (hunk 6 above)
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/workflows/large_scale_docking.py:161]: Yes good spotting, has been fixed upstream.

**Comment by hmacdope** [comment]
Thanks all, @apayne97 should be good for a re-review.

**Comment by hmacdope** [comment]
Dismissing @apayne97 review as addressed and needs to go in :)
